discord_bots/models.py

```diff
@@ -205,6 +206,177 @@ class DiscordMember:
     )
 
 
+@mapper_registry.mapped
+@dataclass
+class EconomyDonation:
+    """
+    Stores economic donation information
+    - Transfers of currency between players
+    """
+
+    __sa_dataclass_metadata_key__ = "sa"
+    __tablename__ = "economy_donation"
+
+    id: str = field(
+        init=False,
+        default_factory=lambda: str(uuid4()),
+        metadata={"sa": Column(String, primary_key=True)},
+    )
+    sending_player_id: int = field(
+        metadata={
+            "sa": Column(BigInteger, ForeignKey("player.id"), nullable=True, index=True)
```

Comment:
Sending player id is null as during `addcurrency` a donation record is created with a null id & the name "Admin". Null ID was chosen over a fake if (ie. 00000000), to avoid fk_constraint errors in pgsql.

This records admin manual additions of currency, while not tying it to that admin (and therefore looking like a legitimate donation). `addcurrency` spawns currency from nowhere, but to maintain transactional integrity, I still needed a source (the donation), as the two-sided transaction.

Reply:
i see. i think what makes this confusing then is we're storing admin donations and transfers between players in the same table, and trying to make them fit in the same abstraction.

they should be either be separated so we can record the admin that created the "admin donation" or we should explicitly store some enum type on economy_donation that distinguishes between admin donations and player-player donations

Reply:
Added `admin_player_id` as a nullable field in EconomyDonation.
So an `addcurrency` would create a donation with no sending player, but an admin player instead.